File: bridge/resources/android/TellingFunctions.kt
package app.lemonfiber.native

import android.Manifest
import android.app.Activity
import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeError
import com.nativephp.mobile.bridge.BridgeFunction

/**
 * Telling somebody something, on Android.
 *
 * Namespace: `Lemonfiber.Telling.*`
 *
 * Local notifications, never pushed. A pushed payload travels through Google's
 * relay to reach this handset, which is a third party reading what a stack said
 * about somebody's home; a local notification is composed here, displayed here,
 * and never leaves the device.
 *
 * The decisions worth testing are not in this file. [NotificationRule] holds
 * what the four facts a platform reports mean together, and [Recurrence] holds
 * when a repeat next comes round; both run on a JVM with no device in sight.
 * What is here is the part that cannot be tested off a handset: reading the
 * platform's facts, posting a notification, arming an alarm.
 *
 * **Nothing rendered is logged.** The title and the body are the only place a
 * stack's name could reach a log line, so the log carries the identifier the
 * caller chose and the outcome word and nothing else. Both of those are closed
 * words rather than rendered text, which is what makes the promise checkable by
 * reading the file.
 */
public object TellingFunctions {
    /** What this plugin's log lines are tagged with. */
    private const val TAG = "Lemonfiber"

    /** Where the record of what this application has asked for is kept. */
    private const val STORE = "lemonfiber.telling"

    /** Whether the notification prompt has ever been raised by this application. */
    private const val EVER_ASKED = "ever_asked"

    /** The identifiers this application has an alarm armed for. */
    private const val ARMED = "armed"

    /** The channel every notification this application posts belongs to. */
    private const val CHANNEL = "lemonfiber.alerts"

    /** What an alarm from this plugin is recognised by. */
    private const val WHEN_IT_IS_TIME = "app.lemonfiber.native.TELL"

    /** The code the permission result is reported back under. */
    private const val ASKING = 0x1EA0

    /** Milliseconds to the second, where the wire counts in seconds. */
    private const val MILLISECONDS = 1_000L

    /**
     * What the platform reports, gathered into the four facts the rule reads.
     *
     * The version check is here rather than inside the rule on purpose: which
     * Android this is, is a fact about the platform, and the rule is the part
     * that has to be runnable without one.
     *
     * `areNotificationsEnabled()` is one call covering three switches. From
     * Android 33 it is false while `POST_NOTIFICATIONS` is not granted, and on
     * every version it is false where the operator has switched the
     * application off in settings — which is the same answer to a caller with
     * something to show, and the reason the rule takes one fact rather than
     * three.
     */
    private fun rule(activity: Activity): NotificationRule =
        NotificationRule(
            wouldAppear = NotificationManagerCompat.from(activity).areNotificationsEnabled(),
            permissionIsAsked = Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU,
            wouldExplain =
                ActivityCompat.shouldShowRequestPermissionRationale(
                    activity,
                    Manifest.permission.POST_NOTIFICATIONS,
                ),
            everAsked = settings(activity).getBoolean(EVER_ASKED, false),
        )

    /** Where this plugin keeps what only it can know. */
    private fun settings(context: Context): SharedPreferences =
        context.getSharedPreferences(STORE, Context.MODE_PRIVATE)

    /**
     * The platform's notification service, which several calls here need.
     *
     * A property rather than a method because that is what it is — the service
     * a context has, not something this object computes.
     */
    private val Context.notifications: NotificationManager
        get() = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    /**
     * The channel every notification here belongs to, created if it is not there.
     *
     * Answers false where the operator has switched that channel off, which is
     * a different refusal from switching the application off: one is answered
     * by sending them to the channel's own settings and the other to the
     * application's.
     *
     * The channel is named after the application, read from the platform rather
     * than written here — the label is already translated into whatever the
     * handset is set to, and a string in this file would be English on a Dutch
     * phone.
     */
    private fun channelIsOpen(context: Context): Boolean {
        val service = context.notifications

        if (service.getNotificationChannel(CHANNEL) == null) {
            service.createNotificationChannel(
                NotificationChannel(
                    CHANNEL,
                    context.applicationInfo.loadLabel(context.packageManager),
                    NotificationManager.IMPORTANCE_DEFAULT,
                ),
            )
        }

        return service.getNotificationChannel(CHANNEL)?.importance != NotificationManager.IMPORTANCE_NONE
    }

    /**
     * One identifier, as a number the platform can key an alarm and a notification by.
     *
     * The sign is taken off because a negative request code is refused by
     * `PendingIntent`.
     */
    private fun codeFor(id: String): Int = id.hashCode() and Int.MAX_VALUE

    /** Post one notification now, or say why it was not posted. */
    private fun post(
        context: Context,
        said: WhatToSay,
    ): Map<String, Any> =
        try {
            context.notifications.notify(
                codeFor(said.id),
                NotificationCompat.Builder(context, CHANNEL)
                    .setContentTitle(said.title)
                    .setContentText(said.body)
                    .setSmallIcon(context.applicationInfo.icon)
                    .setAutoCancel(true)
                    .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                    .build(),
            )

            Log.d(TAG, "telling ${said.id}: shown")
            Envelope.of("shown").asAnswer()
        } catch (refused: SecurityException) {
            // The exception's own message is dropped rather than logged. A
            // platform exception carries what it failed on, and what it failed
            // on here is the rendered sentence.
            Log.w(TAG, "telling ${said.id}: the device refused it (${refused.javaClass.simpleName})")
            Envelope.refusing("withheld", because = "the_device_refused").asAnswer()
        }

    /** The alarm that will post one notification later. */
    private fun pendingFor(
        context: Context,
        said: WhatToSay,
        repeat: Bundle,
    ): PendingIntent =
        PendingIntent.getBroadcast(
            context,
            codeFor(said.id),
            Intent(context, WhenTheTimeComes::class.java)
                .setAction(WHEN_IT_IS_TIME)
                .putExtra("id", said.id)
                .putExtra("title", said.title)
                .putExtra("body", said.body)
                .putExtras(repeat),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )

    /**
     * Arm an alarm, and remember that it is armed.
     *
     * Exact where the platform allows it and approximate where it does not.
     * `USE_EXACT_ALARM` is the permission that would make it always exact, and
     * it is one a store reviews and grants to applications whose whole purpose
     * is an alarm clock. This is not one of those, so a notification that may
     * arrive a few minutes late is the right trade and is made here rather than
     * left to whatever the device happens to allow.
     */
    private fun arm(
        context: Context,
        said: WhatToSay,
        at: Long,
        repeat: Bundle,
    ): Map<String, Any> {
        val alarms = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = pendingFor(context, said, repeat)

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S || alarms.canScheduleExactAlarms()) {
            alarms.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, at, intent)
        } else {
            alarms.set(AlarmManager.RTC_WAKEUP, at, intent)
        }

        remember(context, said.id, armed = true)
        Log.d(TAG, "telling ${said.id}: scheduled")

        return Envelope.of("scheduled").asAnswer()
    }

    /** Add an identifier to the armed set, or take it out of one. */
    private fun remember(
        context: Context,
        id: String,
        armed: Boolean,
    ) {
        val held = armedNow(context).toMutableSet()

        if (armed) held.add(id) else held.remove(id)

        settings(context).edit().putStringSet(ARMED, held).apply()
    }

    /** Every identifier this application has an alarm armed for. */
    private fun armedNow(context: Context): Set<String> =
        settings(context).getStringSet(ARMED, emptySet()) ?: emptySet()

    /** Take one notification away, whether it is showing, armed or neither. */
    private fun withdraw(
        context: Context,
        id: String,
    ) {
        (context.getSystemService(Context.ALARM_SERVICE) as AlarmManager)
            .cancel(pendingFor(context, WhatToSay(id, "", ""), Bundle.EMPTY))

        context.notifications.cancel(codeFor(id))
        remember(context, id, armed = false)
    }

    /**
     * `Lemonfiber.Telling.Standing` — what the operator has already said.
     *
     * Reads and never prompts, which is the whole reason it is a separate
     * function from [Ask]: an application that cannot read the standing answer
     * without raising a dialog has no way to obey one, because reading becomes
     * asking and somebody who already refused gets asked again.
     */
    public class Standing(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> =
            Envelope.of(rule(activity).said.word).asAnswer()
    }

    /**
     * `Lemonfiber.Telling.Ask` — raise the prompt, and record that it was raised.
     *
     * The recording is the part nothing else can do. Android reports *never
     * asked* and *refused for good* identically, so the only way to tell them
     * apart afterwards is to have written down that the dialog went up.
     *
     * It waits for the answer rather than returning while the dialog is still
     * on screen. Returning early would mean reading that record back before the
     * operator had answered — which says *refused* about somebody who is
     * looking at the question. The wait ends when this activity is resumed,
     * which is what happens when the platform's own dialog goes away, and it
     * gives up after two minutes rather than holding the caller for ever.
     */
    public class Ask(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            if (!rule(activity).mayAsk) {
                return Envelope.of(rule(activity).said.word).asAnswer()
            }

            settings(activity).edit().putBoolean(EVER_ASKED, true).apply()

            askFor(activity, Manifest.permission.POST_NOTIFICATIONS, ASKING)

            Log.d(TAG, "telling: the prompt was raised")

            return Envelope.of(rule(activity).said.word).asAnswer()
        }
    }

    /** `Lemonfiber.Telling.Show` — put something in front of the operator now. */
    public class Show(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val said = WhatToSay.from(parameters)

            return when {
                !rule(activity).mayShow -> Envelope.refusing("withheld", "not_permitted").asAnswer()
                !channelIsOpen(activity) -> Envelope.refusing("withheld", "no_such_channel").asAnswer()
                else -> post(activity, said)
            }
        }
    }

    /** `Lemonfiber.Telling.Schedule` — put something in front of them at a time. */
    public class Schedule(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val said = WhatToSay.from(parameters)
            val at =
                (parameters["at"] as? Number)?.toLong()
                    ?: throw BridgeError.InvalidParameters("at is required")

            return when {
                !rule(activity).mayShow -> Envelope.refusing("withheld", "not_permitted").asAnswer()
                !channelIsOpen(activity) -> Envelope.refusing("withheld", "no_such_channel").asAnswer()
                at * MILLISECONDS <= System.currentTimeMillis() ->
                    Envelope.refusing("withheld", "the_time_has_passed").asAnswer()

                else -> arm(activity, said, at * MILLISECONDS, Bundle.EMPTY)
            }
        }
    }

    /** `Lemonfiber.Telling.ScheduleRecurring` — the same, over and over. */
    public class ScheduleRecurring(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val said = WhatToSay.from(parameters)
            val repeat = asked(parameters)
            val next = repeat?.nextAfter(System.currentTimeMillis())

            return when {
                !rule(activity).mayShow -> Envelope.refusing("withheld", "not_permitted").asAnswer()
                !channelIsOpen(activity) -> Envelope.refusing("withheld", "no_such_channel").asAnswer()
                repeat == null || next == null -> Envelope.refusing("withheld", "no_such_repeat").asAnswer()
                else -> arm(activity, said, next, repeat.asExtras())
            }
        }

        /** The recurrence the caller asked for, or nothing where it named none. */
        private fun asked(parameters: Map<String, Any>): Recurrence? =
            HowOften.saying(parameters["frequency"] as? String)?.let { frequency ->
                Recurrence(
                    frequency = frequency,
                    hour = (parameters["hour"] as? Number)?.toInt() ?: 0,
                    minute = (parameters["minute"] as? Number)?.toInt() ?: 0,
                    weekday = (parameters["weekday"] as? Number)?.toInt() ?: 0,
                    dayOfMonth = (parameters["dayOfMonth"] as? Number)?.toInt() ?: 1,
                    month = (parameters["month"] as? Number)?.toInt() ?: 1,
                )
            }
    }

    /**
     * `Lemonfiber.Telling.Cancel` — take one back.
     *
     * Cancelling something that was never scheduled answers `cancelled`, for
     * the reason forgetting a key that was never kept answers `forgotten`: it
     * is the ordinary case after a refusal, and getting rid of something is the
     * one operation that must always work.
     */
    public class Cancel(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            withdraw(activity, WhatToSay.identifierIn(parameters))

            return Envelope.of("cancelled").asAnswer()
        }
    }

    /** `Lemonfiber.Telling.CancelAll` — take back everything this application armed. */
    public class CancelAll(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            armedNow(activity).forEach { withdraw(activity, it) }
            activity.notifications.cancelAll()

            return Envelope.of("cancelled").asAnswer()
        }
    }

    /**
     * `Lemonfiber.Telling.Pending` — what is still to come.
     *
     * Read from this plugin's own record rather than from the platform, which
     * keeps none: `AlarmManager` can be asked to arm one and cannot be asked
     * what it holds.
     */
    public class Pending(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> =
            Envelope.of("read", carrying = mapOf("pending" to armedNow(activity).toList())).asAnswer()
    }

    /**
     * `Lemonfiber.Telling.ClearBadge` — take the count off the launcher icon.
     *
     * Android has no badge of its own: what a launcher draws is a count of the
     * notifications this application is showing. So clearing the badge is
     * clearing those, and there is nothing narrower to do.
     */
    public class ClearBadge(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.notifications.cancelAll()

            return Envelope.of("cleared").asAnswer()
        }
    }

    /**
     * What a scheduled notification is posted by when its time arrives.
     *
     * Declared in the plugin's manifest so that the platform can reach it while
     * this application is not running, which is the whole point of scheduling
     * one.
     *
     * A repeat re-arms itself here rather than being registered as a repeating
     * alarm. `setRepeating` is inexact on every version that matters and
     * carries no way to say *this time of day in this time zone*, so a repeat
     * that crossed a daylight-saving boundary would drift by an hour and stay
     * there.
     */
    public class WhenTheTimeComes : BroadcastReceiver() {
        override fun onReceive(
            context: Context,
            intent: Intent,
        ) {
            val said =
                WhatToSay(
                    id = intent.getStringExtra("id") ?: return,
                    title = intent.getStringExtra("title").orEmpty(),
                    body = intent.getStringExtra("body").orEmpty(),
                )

            channelIsOpen(context)
            post(context, said)

            val repeat = again(intent)
            val next = repeat?.nextAfter(System.currentTimeMillis())

            if (repeat == null || next == null) {
                remember(context, said.id, armed = false)
            } else {
                arm(context, said, next, repeat.asExtras())
            }
        }

        /** The recurrence this alarm carried, or nothing where it was a one-off. */
        private fun again(intent: Intent): Recurrence? =
            HowOften.saying(intent.getStringExtra("frequency"))?.let { frequency ->
                Recurrence(
                    frequency = frequency,
                    hour = intent.getIntExtra("hour", 0),
                    minute = intent.getIntExtra("minute", 0),
                    weekday = intent.getIntExtra("weekday", 0),
                    dayOfMonth = intent.getIntExtra("dayOfMonth", 1),
                    month = intent.getIntExtra("month", 1),
                )
            }
    }
}

/**
 * What one notification says, as this bridge is willing to carry it.
 *
 * Three fields and no more. The vendor's own takes a sound, a badge, a
 * subtitle, an arbitrary data payload and up to three action buttons; none of
 * that has a caller here, and every one of them is another place a value could
 * travel that the application never meant to send.
 */
public data class WhatToSay(
    /** What this notification is about, chosen by the caller and never rendered. */
    public val id: String,
    /** The line the operator reads first. */
    public val title: String,
    /** The line under it. */
    public val body: String,
) {
    /** Where a notification is read out of what the bridge was handed. */
    public companion object {
        /**
         * The three fields, or a refusal naming all of them at once.
         *
         * One guard rather than three, because a caller handed "id is required"
         * learns as much from "id, title and body are required" and the reading
         * stays short enough to check at a glance.
         */
        public fun from(parameters: Map<String, Any>): WhatToSay {
            val id = parameters["id"] as? String
            val title = parameters["title"] as? String
            val body = parameters["body"] as? String

            if (id == null || title == null || body == null) {
                throw BridgeError.InvalidParameters("id, title and body are required")
            }

            return WhatToSay(id, title, body)
        }

        /** The identifier alone, for the calls that carry nothing else. */
        public fun identifierIn(parameters: Map<String, Any>): String =
            parameters["id"] as? String ?: throw BridgeError.InvalidParameters("id is required")
    }
}

/**
 * A recurrence, as an alarm can carry it and hand it back.
 *
 * An extension rather than a method, so that [Recurrence] itself stays free of
 * anything Android: it is a rule, it is unit-tested, and the harness that runs
 * it has no `Bundle` on its classpath.
 */
public fun Recurrence.asExtras(): Bundle =
    Bundle().apply {
        putString("frequency", frequency.word)
        putInt("hour", hour)
        putInt("minute", minute)
        putInt("weekday", weekday)
        putInt("dayOfMonth", dayOfMonth)
        putInt("month", month)
    }

File: bridge/resources/ios/TellingFunctions.swift
import Foundation
import UIKit
import UserNotifications

/// Telling somebody something, on iOS.
///
/// Namespace: `Lemonfiber.Telling.*`
///
/// Local notifications, never pushed. A pushed payload travels through Apple's
/// relay to reach this phone, which is a third party reading what a stack said
/// about somebody's home; a local notification is composed here, displayed
/// here, and never leaves the device.
///
/// The decisions worth testing are not in this file. `NotificationRule` holds
/// what the four facts a platform reports mean together, and `Recurrence` holds
/// when a repeat next comes round; both run on a laptop with no device in
/// sight. What is here is the part that cannot be tested off a phone: reading
/// the platform's facts, adding a request, taking one back.
///
/// **Everything here is synchronous over an asynchronous platform.** A bridge
/// function answers a value, and every `UNUserNotificationCenter` call answers
/// a callback, so each one is waited on. That is what lets `Standing` be a
/// reading rather than a subscription — and a reading is what `Ask` being a
/// separate function depends on.
///
/// **Nothing rendered is logged.** The title and the body are the only place a
/// stack's name could reach a log line, so the log carries the identifier the
/// caller chose and the outcome word and nothing else.
enum TellingFunctions {
    /// Whether the notification prompt has ever been raised by this application.
    private static let everAsked = "lemonfiber.telling.ever_asked"

    /// How long any one platform call is waited on before it is given up on.
    ///
    /// A bridge call that waited for ever would hold the thread it arrived on
    /// for ever. Every wait here is over a callback the system posts promptly,
    /// except the prompt itself — which is an operator reading a dialog, and is
    /// given two minutes.
    private static let patience: TimeInterval = 5

    /// How long the operator is given to answer the prompt.
    private static let patienceWithAPerson: TimeInterval = 120

    /// What the platform reports, gathered into the four facts the rule reads.
    ///
    /// iOS answers this question directly: `UNAuthorizationStatus` has a
    /// `.notDetermined` of its own, so nothing has to be reconstructed. The
    /// rule is asked anyway, with the same four facts, because a rule that
    /// exists on one platform only cannot be seen to disagree with the other.
    ///
    /// `everAsked` is read from this application's own record *or* from the
    /// platform having an answer at all. The record alone would miss somebody
    /// who refused before this version was installed; the platform alone is
    /// what Android cannot supply. Taking either is what makes the two halves
    /// answer the same way.
    private static func rule() -> NotificationRule {
        guard let status = authorizationStatus() else {
            // A centre this process could not read within five seconds. Read as
            // nobody having been asked, which is the safe answer in both
            // directions: it shows nothing, and it records no refusal that
            // nobody made.
            return NotificationRule.unasked
        }

        return NotificationRule(
            wouldAppear: [.authorized, .provisional, .ephemeral].contains(status),
            permissionIsAsked: true,
            wouldExplain: false,
            everAsked: UserDefaults.standard.bool(forKey: everAsked) || status != .notDetermined
        )
    }

    /// What the notification centre has been told, read rather than asked for.
    private static func authorizationStatus() -> UNAuthorizationStatus? {
        var answered: UNAuthorizationStatus?
        let waiting = DispatchSemaphore(value: 0)

        UNUserNotificationCenter.current().getNotificationSettings { settings in
            answered = settings.authorizationStatus
            waiting.signal()
        }

        _ = waiting.wait(timeout: .now() + patience)

        return answered
    }

    /// Hand one request to the centre, and say what became of it.
    private static func add(_ request: UNNotificationRequest, as outcome: String) -> [String: Any] {
        var refused: Error?
        let waiting = DispatchSemaphore(value: 0)

        UNUserNotificationCenter.current().add(request) { error in
            refused = error
            waiting.signal()
        }

        _ = waiting.wait(timeout: .now() + patience)

        if refused != nil {
            // The error's own message is dropped rather than logged. A platform
            // error carries what it failed on, and what it failed on here is
            // the rendered sentence.
            NSLog("%@", "Lemonfiber: telling \(request.identifier): the device refused it")

            return Envelope.refusing("withheld", because: "the_device_refused").asAnswer()
        }

        NSLog("%@", "Lemonfiber: telling \(request.identifier): \(outcome)")

        return Envelope.of(outcome).asAnswer()
    }

    /// What this bridge is willing to put in a notification, and nothing else.
    private static func content(_ said: WhatToSay) -> UNMutableNotificationContent {
        let content = UNMutableNotificationContent()

        content.title = said.title
        content.body = said.body

        return content
    }

    /// `Lemonfiber.Telling.Standing` — what the operator has already said.
    ///
    /// Reads and never prompts, which is the whole reason it is a separate
    /// function from `Ask`: an application that cannot read the standing answer
    /// without raising a dialog has no way to obey one, because reading becomes
    /// asking and somebody who already refused gets asked again.
    class Standing: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            BridgeResponse.success(data: Envelope.of(rule().said.word).asAnswer())
        }
    }

    /// `Lemonfiber.Telling.Ask` — raise the prompt, and record that it was raised.
    ///
    /// The recording is what keeps the two platforms answering alike. iOS does
    /// not need it — it reports `.notDetermined` itself — and Android cannot
    /// answer without it, so it is written on both and read on both.
    ///
    /// It waits for the operator rather than returning while the dialog is
    /// still on screen, so that the answer it gives is the answer they gave.
    class Ask: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard rule().mayAsk else {
                return BridgeResponse.success(data: Envelope.of(rule().said.word).asAnswer())
            }

            UserDefaults.standard.set(true, forKey: everAsked)

            let waiting = DispatchSemaphore(value: 0)

            let centre = UNUserNotificationCenter.current()

            centre.requestAuthorization(options: [.alert, .sound, .badge]) { _, _ in
                waiting.signal()
            }

            _ = waiting.wait(timeout: .now() + patienceWithAPerson)

            NSLog("%@", "Lemonfiber: telling: the prompt was raised")

            return BridgeResponse.success(data: Envelope.of(rule().said.word).asAnswer())
        }
    }

    /// `Lemonfiber.Telling.Show` — put something in front of the operator now.
    ///
    /// There is no `no_such_channel` here. Channels are Android's; iOS has one
    /// switch per application and the rule already reads it, so the outcome
    /// exists on the wire and is never the answer on this platform.
    class Show: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let said = WhatToSay.from(parameters) else {
                return BridgeResponse.error(
                    code: "INVALID_PARAMETERS",
                    message: "id, title and body are required"
                )
            }

            guard rule().mayShow else {
                return BridgeResponse.success(
                    data: Envelope.refusing("withheld", because: "not_permitted").asAnswer()
                )
            }

            return BridgeResponse.success(
                data: add(
                    UNNotificationRequest(identifier: said.id, content: content(said), trigger: nil),
                    as: "shown"
                )
            )
        }
    }

    /// `Lemonfiber.Telling.Schedule` — put something in front of them at a time.
    class Schedule: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let said = WhatToSay.from(parameters), let at = parameters["at"] as? Int else {
                return BridgeResponse.error(
                    code: "INVALID_PARAMETERS",
                    message: "id, title, body and at are required"
                )
            }

            let when = Date(timeIntervalSince1970: TimeInterval(at))

            guard rule().mayShow else {
                return BridgeResponse.success(
                    data: Envelope.refusing("withheld", because: "not_permitted").asAnswer()
                )
            }

            guard when > Date() else {
                return BridgeResponse.success(
                    data: Envelope.refusing("withheld", because: "the_time_has_passed").asAnswer()
                )
            }

            let fields = Calendar.current.dateComponents(
                [.year, .month, .day, .hour, .minute, .second],
                from: when
            )

            return BridgeResponse.success(
                data: add(
                    UNNotificationRequest(
                        identifier: said.id,
                        content: content(said),
                        trigger: UNCalendarNotificationTrigger(dateMatching: fields, repeats: false)
                    ),
                    as: "scheduled"
                )
            )
        }
    }

    /// `Lemonfiber.Telling.ScheduleRecurring` — the same, over and over.
    ///
    /// The platform repeats this one itself, which Android's alarms cannot do
    /// across a time-zone change — so there the bridge re-arms each occurrence
    /// by hand and here it hands the fields over once. The fields are the
    /// rule's, on both, which is what keeps the two from disagreeing about what
    /// `weekly` means.
    class ScheduleRecurring: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let said = WhatToSay.from(parameters) else {
                return BridgeResponse.error(
                    code: "INVALID_PARAMETERS",
                    message: "id, title and body are required"
                )
            }

            guard rule().mayShow else {
                return BridgeResponse.success(
                    data: Envelope.refusing("withheld", because: "not_permitted").asAnswer()
                )
            }

            guard let repeating = asked(parameters), repeating.nextAfter(Date()) != nil else {
                return BridgeResponse.success(
                    data: Envelope.refusing("withheld", because: "no_such_repeat").asAnswer()
                )
            }

            return BridgeResponse.success(
                data: add(
                    UNNotificationRequest(
                        identifier: said.id,
                        content: content(said),
                        trigger: UNCalendarNotificationTrigger(
                            dateMatching: repeating.fixes().asDateComponents(),
                            repeats: true
                        )
                    ),
                    as: "scheduled"
                )
            )
        }

        /// The recurrence the caller asked for, or nothing where it named none.
        private func asked(_ parameters: [String: Any]) -> Recurrence? {
            guard let frequency = HowOften.saying(parameters["frequency"] as? String) else {
                return nil
            }

            return Recurrence(
                frequency: frequency,
                hour: parameters["hour"] as? Int ?? 0,
                minute: parameters["minute"] as? Int ?? 0,
                weekday: parameters["weekday"] as? Int ?? 0,
                dayOfMonth: parameters["dayOfMonth"] as? Int ?? 1,
                month: parameters["month"] as? Int ?? 1
            )
        }
    }

    /// `Lemonfiber.Telling.Cancel` — take one back.
    ///
    /// Cancelling something that was never scheduled answers `cancelled`, for
    /// the reason forgetting a key that was never kept answers `forgotten`: it
    /// is the ordinary case after a refusal, and getting rid of something is
    /// the one operation that must always work.
    class Cancel: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let id = parameters["id"] as? String else {
                return BridgeResponse.error(code: "INVALID_PARAMETERS", message: "id is required")
            }

            let centre = UNUserNotificationCenter.current()

            centre.removePendingNotificationRequests(withIdentifiers: [id])
            centre.removeDeliveredNotifications(withIdentifiers: [id])

            return BridgeResponse.success(data: Envelope.of("cancelled").asAnswer())
        }
    }

    /// `Lemonfiber.Telling.CancelAll` — take back everything this application armed.
    class CancelAll: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let centre = UNUserNotificationCenter.current()

            centre.removeAllPendingNotificationRequests()
            centre.removeAllDeliveredNotifications()

            return BridgeResponse.success(data: Envelope.of("cancelled").asAnswer())
        }
    }

    /// `Lemonfiber.Telling.Pending` — what is still to come.
    ///
    /// Read from the platform, which keeps the list itself. Android's
    /// `AlarmManager` can be asked to arm one and cannot be asked what it
    /// holds, so the bridge keeps its own record there; the answer on the wire
    /// is the same either way, which is the point of the wire.
    class Pending: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            var identifiers: [String] = []
            let waiting = DispatchSemaphore(value: 0)

            UNUserNotificationCenter.current().getPendingNotificationRequests { requests in
                identifiers = requests.map(\.identifier)
                waiting.signal()
            }

            _ = waiting.wait(timeout: .now() + patience)

            return BridgeResponse.success(
                data: Envelope.of("read", carrying: ["pending": identifiers]).asAnswer()
            )
        }
    }

    /// `Lemonfiber.Telling.ClearBadge` — take the count off the home-screen icon.
    ///
    /// A real badge here, unlike Android, where what a launcher draws is a
    /// count of the notifications showing and clearing it means clearing those.
    class ClearBadge: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            if #available(iOS 16.0, *) {
                UNUserNotificationCenter.current().setBadgeCount(0)
            } else {
                DispatchQueue.main.async {
                    UIApplication.shared.applicationIconBadgeNumber = 0
                }
            }

            return BridgeResponse.success(data: Envelope.of("cleared").asAnswer())
        }
    }
}

/// What one notification says, as this bridge is willing to carry it.
///
/// Three fields and no more. The vendor's own takes a sound, a badge, a
/// subtitle, an arbitrary data payload and up to three action buttons; none of
/// that has a caller here, and every one of them is another place a value could
/// travel that the application never meant to send.
struct WhatToSay {
    /// What this notification is about, chosen by the caller and never rendered.
    let id: String

    /// The line the operator reads first.
    let title: String

    /// The line under it.
    let body: String

    /// The three fields, or a refusal naming all of them at once.
    ///
    /// One guard rather than three, because a caller handed "id is required"
    /// learns as much from "id, title and body are required" and the reading
    /// stays short enough to check at a glance.
    ///
    /// - Parameter parameters: what the bridge was handed.
    /// - Returns: what to say, or nil where any of the three is missing.
    static func from(_ parameters: [String: Any]) -> WhatToSay? {
        guard
            let id = parameters["id"] as? String,
            let title = parameters["title"] as? String,
            let body = parameters["body"] as? String
        else {
            return nil
        }

        return WhatToSay(id: id, title: title, body: body)
    }
}
